pkg/sparrow/gitlab/gitlab.go

```diff
@@ -0,0 +1,311 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package gitlab
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/caas-team/sparrow/internal/logger"
+	"github.com/caas-team/sparrow/pkg/checks"
+)
+
+// Gitlab handles interaction with a gitlab repository containing
+// the global targets for the Sparrow instance
+type Gitlab interface {
+	FetchFiles(ctx context.Context) ([]checks.GlobalTarget, error)
+	PutFile(ctx context.Context, file File) error
+	PostFile(ctx context.Context, file File) error
+}
+
+// Client implements Gitlab
+type Client struct {
+	// the base URL of the gitlab instance
+	baseUrl string
+	// the ID of the project containing the global targets
+	projectID int
+	// the token used to authenticate with the gitlab instance
+	token  string
+	client *http.Client
+}
+
+func New(baseURL, token string, pid int) Gitlab {
+	return &Client{
+		baseUrl:   baseURL,
+		token:     token,
+		projectID: pid,
+		client:    &http.Client{},
+	}
+}
+
+// FetchFiles fetches the files from the global targets repository from the configured gitlab repository
+func (g *Client) FetchFiles(ctx context.Context) ([]checks.GlobalTarget, error) {
+	log := logger.FromContext(ctx)
+	fl, err := g.fetchFileList(ctx)
+	if err != nil {
+		log.Error("Failed to fetch files", "error", err)
+		return nil, err
+	}
+
+	var result []checks.GlobalTarget
+	for _, f := range fl {
+		gl, err := g.fetchFile(ctx, f)
+		if err != nil {
+			log.Error("Failed fetching files", "error", err)
+			return nil, err
+		}
+		result = append(result, gl)
+	}
+	log.Info("Successfully fetched all target files", "files", len(result))
+	return result, nil
+}
+
+// fetchFile fetches the file from the global targets repository from the configured gitlab repository
+func (g *Client) fetchFile(ctx context.Context, f string) (checks.GlobalTarget, error) {
+	log := logger.FromContext(ctx)
+	var res checks.GlobalTarget
+	// URL encode the name
+	n := url.PathEscape(f)
+	req, err := http.NewRequestWithContext(ctx,
+		http.MethodGet,
+		fmt.Sprintf("%s/api/v4/projects/%d/repository/files/%s/raw?ref=main", g.baseUrl, g.projectID, n),
+		http.NoBody,
+	)
+	if err != nil {
+		log.Error("Failed to create request", "error", err)
+		return res, err
+	}
+	req.Header.Add("PRIVATE-TOKEN", g.token)
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := g.client.Do(req) //nolint:bodyclose // closed in defer
+	if err != nil {
+		log.Error("Failed to fetch file", "file", f, "error", err)
+		return res, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		log.Error("Failed to fetch file", "status", resp.Status)
+		return res, fmt.Errorf("request failed, status is %s", resp.Status)
+	}
+
+	defer func(Body io.ReadCloser) {
```

Comment:
Might be better to add the defer before the status check, doesn't it?

The body should be closed even if the code is an error code.